First test complete

TheSprtCz opened the pull request:
# Overview
First test complete. I used Spock framework, which is contained in Grails, but sadly only 2.3+, so for 2.2 there needs to be plugin. Also I added testing environment.

# Changes in files:
*BootStrap.groovy*
- added testing environment with some user, applications, topics etc.

*BuildConfig.groovy*
- added Spock plugin dependency
- added http://repo.grails.org/grails/repo/ as a repo, because without that i wasn't able to download all dependencies

*UniversityControllerIntegrationSpec.groovy*
- Integration test for UniversityController
- all methods tested, except for update

VaclavDedik commented:
Great! Just one major problem, convert the tabs to spaces (4 spaces) and configure your editor to use spaces. Tabs are awful :)

VaclavDedik commented:
Also, consider creating a new branch for each feature/issue you work on. Putting everything in your master isn't a good idea.

TheSprtCz commented:
Ok, all tabs should be removed.

jcechace commented:
I would suggest squashing the commits. Commits which are just fixing formatting issues introduced by directly preceding commits do not have any real value.

TheSprtCz commented:
Ok, squashed

jcechace commented:
TBH I don't quite understand the reason for these formatting changes in BootStrap.groovy.

jcechace commented on test/integration/com/redhat/theses/UniversityControllerIntegrationSpec.groovy:
```diff
@@ -0,0 +1,131 @@
+        model.universityInstance != null
+        model.universityInstance.name == "First"
+        model.universityInstance.acronym == "F"
+    }
```
* Names such as "test show() true" are not really suitable names.
* Names such as "can retrieve university by id" are
* Empty lines in between test cases/methods would help improve the overall readability of test files. 

Compare your version with my, slightly modified one:
```groovy
void "can retrieve university"()  {
    when: "retrieving university by used id"
        def model = controller.show(1)
    then: "should retrieve the correct entity "
        model?.universityInstance?.name == "First"
        model?.universityInstance?.acronym == "F"
 }
```

BTW: ?. is a handy "safe" method call. Attribute/method will be called if it is defined on given object, otherwise null will be returned.

TheSprtCz commented:
I pushed new commit, most of the issues should be solved now.
* Made tests more readable.
* Removed test on create() because it has no fail-state and parameters don't change output of the function.
* Removed duplicate test on edit().
* Small changes to the testing environment
* Reverted back most of the formatting changes to **BootStrap.groovy**. Changes were not intentional, I used an IDE to change tabs for space and it also changed formatting in the whole file, should be ok now.

jcechace commented:
Once again squash those commits!

jcechace commented:
I can still see a bunch of pointless formatting changes. I am not against reformatting the code per se, however such thing should be done in separate commit. Mixing it in with some other changes makes git blame and potential revers needlessly complicated.

jcechace commented on test/integration/com/redhat/theses/UniversityControllerIntegrationSpec.groovy:
```diff
@@ -0,0 +1,111 @@
+            controller.response?.redirectUrl.endsWith("list")
+    }
+    
+    void "Can edit university, should fail"() {
```
Not like this..  The point of this BDD style naming is to use plain English as much as possible.

"Should not be able to edit university" 

or even better

"Can not edit university as regular user"

jcechace commented:
I am also not a fan of having all this test configuration as part of BootStrap directly. However I guess we can keep it if there is no other way.

TheSprtCz commented:
* I updated test names
* Found way to separate test environment to its own file (**TestBootStrap.groovy**)
  * **BootStrap.groovy** should be now completely unchanged
* Changed repo from http://repo.grails.org/grails/core/ to http://repo.grails.org/grails/plugins/ because for some unknown reason it just stopped working.

jcechace commented on grails-app/conf/BuildConfig.groovy:
```diff
@@ -32,18 +32,18 @@ grails.project.dependency.resolution = {
         // runtime 'mysql:mysql-connector-java:5.1.20'
         runtime 'postgresql:postgresql:8.4-702.jdbc4'
-
+		test "org.spockframework:spock-grails-support:0.7-groovy-2.0"
```
incorrect indentation?

TheSprtCz replied:
Don't know why it was Github showing like this, when I went into repo and opened that file or looked at raw version, it looked normal, without any indentation problems. Anyway I moved that line to different spot and it looks better now.

VaclavDedik commented:
Merged, thanks!
